src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -344,8 +425,20 @@ ExecuteTaskAndStoreResults(QueryDesc *queryDesc, Task *task,
 		 */
 		if (!gotResults && expectResults)
 		{
-			queryOK = StoreQueryResult(routerState, connection, tupleDescriptor,
-									   &currentAffectedTupleCount);
+			/* use try/catch to maintain txn participant list during fail-fast errors */
+			PG_TRY();
```

Comment:
Hm. That pretty much assumes the error is a forwarded remote error. What if some memory allocation or such failed instead? We'll leave the connection open afaics. I'm a bit inclined towards stopping purging such connections instead.